_posts/2021-11-19-connecting-java-high-level-rest-client-with-opensearch-over-https.markdown

```diff
@@ -0,0 +1,176 @@
+
+### STEP 1: Convert the root authority certificates from `.pem` to `.der` format.
+
+```
+openssl x509 -in opensearch-1.1.0/config/root-ca.pem -inform pem -out root-ca.der --outform der
+```
+
+### STEP 2: Create a custom truststore and add the `root-ca.der` certs. 
+
+Adding the root authority certificate to the application truststore tells the application to trust any certificate signed by this root CA.
+
+```
+keytool -import root-ca.der -alias opensearch -keystore myTrustStore
+```
+
+Confirm the action was successful by listing certs in truststore. The `grep` should be able to find opensearch alias if the certs were added successfully.
+
+```
+keytool -keystore myTrustStore -storepass changeit -list | grep opensearch
+```
+
+### STEP 3: Next, set the truststore properties in the Java application code to point to the custom truststore. 
+
+```
+// Point to keystore with appropriate certificates for security.
+System.setProperty("javax.net.ssl.trustStore", "/full/path/to/myCustomTrustStore");
+System.setProperty("javax.net.ssl.trustStorePassword", "password-for-myCustomTrustStore");
+```
+
+
+### STEP 4: Now create an instance of the client and connect to OpenSearch over HTTPS. 
+
+```
+// Create the client.
+RestClientBuilder builder = RestClient.builder(new HttpHost("localhost", 9200, "https"))
+        .setHttpClientConfigCallback(new RestClientBuilder.HttpClientConfigCallback() {
+            @Override
+            public HttpAsyncClientBuilder customizeHttpClient(HttpAsyncClientBuilder httpClientBuilder) {
+                return httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider);
+            }
+        });
+RestHighLevelClient client = new RestHighLevelClient(builder);
+
+// Create an index with custom settings.
+CreateIndexRequest createIndexRequest = new CreateIndexRequest("custom-index");
+createIndexRequest.settings(Settings.builder() //Specify in the settings how many shards you want in the index.
+    .put("index.number_of_shards", 4)
+    .put("index.number_of_replicas", 1)
+    );
+CreateIndexResponse createIndexResponse = client.indices().create(createIndexRequest, RequestOptions.DEFAULT);
+
+// Adding data to the index.
+IndexRequest request = new IndexRequest("custom-index"); //Add a document to the custom-index we created.
+request.id("1"); //Assign an ID to the document.
+HashMap<String, String> stringMapping = new HashMap<String, String>();
+stringMapping.put("message:", "Testing Java REST client");
+request.source(stringMapping); //Place your content into the index's source.
+IndexResponse indexResponse = client.index(request, RequestOptions.DEFAULT);
+```
+
+Checkout this [documentation](https://opensearch.org/docs/latest/clients/java-rest-high-level/) for full sample code.
```

Comment:
Check out

full -> complete

or maybe "working"?

Reply:
I think some output from the run of the sample code would be helpful.

Reply:
I think providing output would have made more sense if I could put out github example repo which readers could directly run and validate with a simple gradle command. In this case, it might cause confusion since I'm not asking the reader to run the sample code, but only pointing to it as one example.